plugins/auth/src/pam.cpp

```diff
@@ -0,0 +1,373 @@
+#include "irods/authentication_plugin_framework.hpp"
+
+#define USE_SSL 1
+#include "irods/sslSockComm.h"
+
+#include "irods/icatHighLevelRoutines.hpp"
+#include "irods/irods_at_scope_exit.hpp"
+#include "irods/irods_auth_constants.hpp"
+#include "irods/irods_client_server_negotiation.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_pam_auth_object.hpp"
+#include "irods/rcConnect.h"
+
+#include <boost/lexical_cast.hpp>
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
+
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <string>
+#include <termios.h>
+#include <unistd.h>
+
+namespace
+{
+    namespace irods_auth = irods::experimental::auth;
+    using json = nlohmann::json;
+
+    auto request(rcComm_t& comm, const json& msg)
+    {
+        auto str = msg.dump();
+
+        bytesBuf_t inp;
+        inp.buf = static_cast<void*>(const_cast<char*>(str.c_str()));
+        inp.len = str.size();
+
+        bytesBuf_t* resp{};
+        auto ec = procApiRequest(&comm, 110000, static_cast<void*>(&inp), nullptr, reinterpret_cast<void**>(&resp), nullptr);
+
+        if (ec < 0) {
+            THROW(ec, "failed to perform request");
+        }
+
+        return json::parse(static_cast<char*>(resp->buf), static_cast<char*>(resp->buf) + resp->len);
+    } // request
+
+    // TODO: std::vector could be a const ref
+    auto verify(const json& p, std::vector<std::string> n)
+    {
+        for(auto&& x : n) {
+            if(!p.contains(x)) {
+                return std::make_tuple(false, x);
+            }
+        }
+
+        return std::make_tuple(true, std::string{});
+    } // verify
```

Comment:
This duplicates a function in the native authentication plugin. Not 100% sure how useful it is, but it feels this could be included in the framework?

Reply:
And here

https://github.com/irods/irods/blob/f6fa01c972cd36a8f56c8caae5cbe82e1041b90e/plugins/auth/src/native.cpp#L38-L47

Reply:
If it is duplicated, we should adjust the code to avoid that where it makes sense.

So yes, I agree.